Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+However, the intrinsics will be added as needed or requested for various optimization efforts.
+
++ The set of intrinsics instructions supported by a namespace should only be allowed
+if all platforms reporting `IsSupported` are supported (at least at time of release)
++ It may be preferable to add new namespaces in cases where this creates significant burden
+
+### Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ ARM64 will introduce `System.Runtime.Intrinsics.Half` to support this request 
++ `Half` will be defined as:
+
+```
+    [StructLayout(LayoutKind.Sequential, Size = 2)]
+    public struct Half : struct {}
+``` 
+
+#### ARMv8 Half precision support
+
+ARMv8 baseline support for `Half` is limited.  The following operations are supported
+
++ Loads and Stores
++ Conversion to/from `Float`
++ Widening from `Vector128<Half>` to two `Vector128<Float>`
++ Narrowing from two `Vector128<Float>` to `Vector128<Float>`
+
+It is presumed that even this minimal support could be helpful
+
+Optional `Half` extensions add more complete support
+
+#### `Half` and ARM64 ABI
+
+The proposed `Half` implementation will treat `Half` as raw bits.  
+It is likely therefore that this will not conform to the ARM64 ABI specifically with respect to the HFA
+calling convention.  This will hinder use of `Half` in `PInvoke` calls.  Resolving this ABI issue will
+be a low priority.
+
+### Scalable Vector Extension Support
+
+`SVE`, the Scalable Vector Extension introduces its own complexity.  
+
+The extension 
++ Creates a new set of SVE registers 
++ Each register has a platform specific length
+  + Any multiple of 128 bits
+
+Therefore implementation will not be trivial.
+
++ Register allocator may need changes
++ Crossgen of SVE intrinsics must be delayed until runtime JIT
++ SIMD support will face similar issues
++ Open issue: Should we use `Vector<T>` or SVE<T> in user interface design?
++ SVE probably requires a separate design
+
+### Miscellaneous
+#### Choice of Arm64 naming conventions -- Rationale
+
+##### Naming conventions
+
++ Namespaces are generally using Pascal case
++ Namespaces and Names must not start with a number
++ Names are descriptive.  Therefore intrinsics are generally named to describe functionality
+
+`ARMv8` is assumed to be an abbreviation for `ARM Reference Manual Version 8`
+
+##### Reference x86 & x64 choice of namespaces
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.X86` is used for enumerations shared across both platforms
++ `System.Runtime.Intrinsics.AVX` is used for the instructions which are included in a specific hardware extension
```

Comment:
There is not namespace called `System.Runtime.Intrinsics.AVX` in the current x86/x64 intrinsics. The whole x86/x64 intrinsics are in a single `System.Runtime.Intrinsics.X86` namespace.

Reply:
Yes.... `System.Runtime.Intrinsics.AVX`, is a typo and a mistake.  `System.Runtime.Intrinsics.X86.AVX` is a class/struct, not a namespace.